File: c++/90. Subsets II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> tmp;
        res.push_back(tmp);
        sort(nums.begin(), nums.end());
        int start, end, p = 0;
        
        for (int i = 0; i < nums.size(); i++) {
            start = (i!=0 && nums[i] == nums[i-1])?p:0;
            end = res.size();
            if (start == end) continue;
            
            for (int j = start; j < end; j++) {
                vector<int> v;
                for (int k = 0; k < res[j].size(); k++)
                    v.push_back(res[j][k]);
                v.push_back(nums[i]);
                res.push_back(v);
            }
            
            p = end;
        }
        
        return res;
    }
};

File: python/90. Subsets II.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        res = [[]]
        nums.sort()
        last = 0
        for i in range(n):
            l = 0
            if i>0 and nums[i-1]==nums[i]:
                l = last
            r = len(res)

            for j in range(l, r):
                res.append(res[j] + [nums[i]])
            last = r
                
        return res

# 思路: last记录上一轮加入到结果中子集，如果这轮的数字是重复的，则从last位置开始生成子集